algorithms/src/snark/marlin/data_structures/proof.rs

```diff
@@ -111,43 +127,42 @@ pub struct WitnessCommitments<E: PairingEngine> {
     pub z_b: sonic_pc::Commitment<E>,
 }
 
+// TODO: we're deserializing this, should we still keep references to objects...?
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Evaluations<F: PrimeField> {
-    /// Evaluation of `z_b_i`'s at `beta`.
-    pub z_b_evals: Vec<F>,
+    /// Evaluations of `z_b_i_j`'s at `beta`.
+    pub z_b_evals: BTreeMap<CircuitId, Vec<F>>,
     /// Evaluation of `g_1` at `beta`.
     pub g_1_eval: F,
-    /// Evaluation of `g_a` at `beta`.
-    pub g_a_eval: F,
-    /// Evaluation of `g_b` at `gamma`.
-    pub g_b_eval: F,
-    /// Evaluation of `g_c` at `gamma`.
-    pub g_c_eval: F,
+    /// Evaluation of `g_a_i`'s at `beta`.
```

Comment:
I'm not sure that we should be serializing the circuit IDs; instead, it should come from the verification key directly.

So I'd change this to `Vec<F>`, and change `z_b_evals` to `Vec<Vec<F>>`

Reply:
Done: d954e90b3

Are you aware of any tests which do serialization so we can test if the ordering is correct?

Reply:
No, but we should add them.

Reply:
48f99f49d